warden_daemon/src/virtualization/qemu_runner.rs

```diff
@@ -0,0 +1,99 @@
+use std::path::PathBuf;
+use std::process::{Child, Command, ExitStatus, Stdio};
+
+use log::trace;
+
+use crate::managers::realm_configuration::{CpuConfig, KernelConfig, MemoryConfig, NetworkConfig};
+use crate::managers::realm_manager::{VmManager, VmManagerError};
+
+pub struct QemuRunner {
+    command: Command,
+    vm: Option<Child>,
+}
+
+impl QemuRunner {
+    pub fn new(path_to_qemu: PathBuf) -> Self {
+        QemuRunner {
+            command: Command::new(path_to_qemu),
+            vm: None,
+        }
+    }
+}
+
+impl VmManager for QemuRunner {
+    fn setup_network(&mut self, config: &NetworkConfig) {
+        self.command.arg("-netdev").arg(format!(
+            "tap,id=mynet0,ifname={},script=no,downscript=no",
+            &config.tap_device
+        ));
+        self.command.arg("-device").arg(format!(
+            "{},netdev=mynet0,mac={}",
+            config
+                .hardware_device
+                .as_ref()
+                .get_or_insert(&String::from("e1000")),
+            config.mac_address
+        ));
+        self.command.arg("-device").arg(format!(
+            "vhost-vsock-pci,id=vhost-vsock-pci0,guest-cid={}",
+            config.vsock_cid
+        ));
+        if let Some(terminal_uri) = &config.remote_terminal_uri {
+            // Setup access terminal
+            self.command.arg("-serial").arg(terminal_uri);
+        }
+    }
+    fn setup_kernel(&mut self, config: &KernelConfig) {
+        self.command.arg("-kernel").arg(&config.kernel_path);
+    }
+    fn setup_cpu(&mut self, config: &CpuConfig) {
+        self.command
+            .arg("-smp")
+            .arg(config.cores_number.to_string());
+        self.command.arg("-cpu").arg(&config.cpu);
+    }
+    fn setup_memory(&mut self, config: &MemoryConfig) {
+        self.command.arg("-m").arg(config.ram_size.to_string());
+    }
+    fn setup_machine(&mut self, name: &str) {
+        self.command.arg("-machine").arg(name);
+    }
+    fn launch_vm(&mut self) -> Result<(), VmManagerError> {
+        self.control_output();
+        trace!("Spawning realm with command: {:?}", self.command);
+        self.command
+            .spawn()
+            .map(|child| {
+                self.vm = Some(child);
+                self.command = Command::new(self.command.get_program());
+            })
+            .map_err(VmManagerError::LaunchFail)
+    }
+    fn stop_vm(&mut self) -> Result<(), VmManagerError> {
+        self.vm
+            .as_mut()
+            .map(|child| child.kill().map_err(|_| VmManagerError::StopFail))
+            .unwrap_or(Err(VmManagerError::VmNotLaunched))
```

Comment:
If I'm correct Child::kill doesn't run waitpid to collect the return code. Shouldn't the stop_vm function do it or do you intend to run get_exit_status somewhere else?

Reply:
You are right, this module is just a very simple runner for qemu and in the future I will rewrite it to use tokio::command but for now I will also add collecting the exit status to make sure that there are no zombie processes. Done